### ISSUE TITLE
JObject does not deserialize double correctly

### ISSUE THREAD

**Issue by RDUser-dev** [JamesNK/Newtonsoft.Json#2889]
### Generate Json from a JObject and load back into a second JObject

```csharp
JObject obj1 = new JObject() { 
                { "val1", -5.2585885403382 } 
            };
var str1 = obj1.ToString();
```
`str1` contains the expected value:
```javascript
{
  "val1": -5.2585885403382
}
```
Try loading the json string into a JObject again
 
```csharp
JObject obj2 = JObject.Parse(str1); 
```
Now `obj2["Val1"]` contains a different value and stringify to a different Json:
```javascript
{
  "val1": -5.2585885403382004
}
```
### Expected behavior

`obj2` equals `obj1`

### Actual behavior

With .netFramework 4.8 running at 64 bit
`obj2` and `obj1` are different

### Steps to reproduce

- Run the code beblow in .netFramework 4.8 compiled with Prefer 32 bit flag enabled, the test pass.
- Run the same code compiled without Prefer 32 bit flag enabled, the test fails.

```csharp
static void Main(string[] args)
        {
            JObject obj1 = new JObject() { 
                { "val1", -5.2585885403382 } 
            }; 

            JObject obj2 = JObject.Parse(obj1.ToString());

            if (obj2["val1"].Value<double>() != -5.2585885403382) throw new ApplicationException();
        }
```

### Notes
We use JObjects to save configurations by attaching a hash of the json, we expect to be able to read and re-save the same configuration getting the same hash. It is therefore important that the serialization and deserialization of the JObject is perfectly symmetric, and not just for my specific scenario.
Do you know if there is an existing workaround to the problem?

**Comment by elgonzo** [comment]
Cannot reproduce with the current Newtonsoft.Json version 13.0.3, using a VS2015-style project targetting .NET Framework 4.8 and the Prefer32Bit flag being enabled. `obj2["val1"].Value<double>()` yields exactly -5.2585885403382.

Make sure you are using the current Newtonsoft.Json 13.0.3. If the problem still persists with 13.0.3, could you please share your project file you are used for running the sample Main method you posted?

**Comment by RDUser-dev** [comment]
> _Edit: I can reproduce with Prefer32Bit being disabled on a x64 machine. Can you please reconfirm that you see the problem when Prefer32Bit is enabled or when it is disabled?_
> 
> Cannot reproduce with the current Newtonsoft.Json version 13.0.3, using a VS2015-style project targetting .NET Framework 4.8 and the Prefer32Bit flag being enabled. `obj2["val1"].Value<double>()` yields exactly -5.2585885403382.
> 
> Make sure you are using the current Newtonsoft.Json 13.0.3. If the problem still persists with 13.0.3, could you please share your project file you are used for running the sample Main method you posted?

Thanks Elgonzo for the quick reply,
I am sending you the test described above.
[TestDoubleNewtonsoft.zip](https://github.com/JamesNK/Newtonsoft.Json/files/12440420/TestDoubleNewtonsoft.zip)

**Comment by elgonzo** [comment]
At least regarding my tests, the problem lies with the compiler. With Prefer32Bit being disabled, the -5.2585885403382 will be turned by the compiler into the effective value -5.2585885403381996. The intended value probably cannot be represented exactly as a floating point value. Interestingly, double.Parse(string) results in a slightly different incorrect value. Demonstration:

```C#
double d1 = -5.2585885403382;
double d2 = -5.2585885403382d;
double d3 = double.Parse("-5.2585885403382");

Console.WriteLine(d1.ToString("G17"));
Console.WriteLine(d2.ToString("G17"));
Console.WriteLine(d3.ToString("G17"));
```
will output:
```
-5.2585885403381996
-5.2585885403381996
-5.2585885403382004
```

Long story short: As is often the case, beware precision-related issues with floating point types. If you need exact value and want to do equality comparisons without accounting for epsilon, i'd suggest you switch to using decimal type instead of double/float. For example like this:
```C#
    JObject obj2 = JObject.Parse(str1);
    decimal d = obj2["val1"].Value<decimal>();

   if (d != -5.2585885403382m) throw new ApplicationException();    
```

**Comment by RDUser-dev** [comment]
> At least regarding my tests, the problem lies with the compiler. With Prefer32Bit being disabled, the -5.2585885403382 will be turned by the compiler into the effective value -5.2585885403381996. I guess, the intended value probably cannot be represented exactly as a double floating point value. Interestingly, double.Parse(string) results in a slightly different incorrect value. 

I thought there were problems with rounding conversions etc..
But even if the strings do not exactly match the starting number, they coincide with each other consistently.
The point is it is possible to do the conversions double=>string=>double=>string and the two strings match.
Instead it is not possible to do the conversions JObject=>JsonString=>JObject=>JsonString and obtains two equal JsonString.

![image](https://github.com/JamesNK/Newtonsoft.Json/assets/49058637/fa9d19d1-d438-4393-a601-881a8d7e2100)

moving from double to decimal would be a huge job, since the code we are working on is very large.
But thanks for the tip

**Comment by elgonzo** [comment]
> Instead it is not possible to do the conversions JObject=>JsonString=>JObject=>JsonString and obtains two equal JsonString.

Why do you think it is not possible to do JObject=>JsonString=>JObject=>JsonString?

Based on your example code, if you do:
```C#
var str1 = obj1.ToString();

JObject obj2 = JObject.Parse(str1);

var str2 = obj2.ToString();
Console.WriteLine(str1 == str2);
```
str2 will be equal to str1.

Do you have a concrete example where JObject=>JsonString=>JObject=>JsonString fails?

**Comment by RDUser-dev** [comment]
Good morning @elgonzo

> 
> Do you have a concrete example where JObject=>JsonString=>JObject=>JsonString fails?

Of course, the first example does pretty much that. Run in 64 bit it fails

```CSharp

        static void Main(string[] args)
        {
            
            JObject obj1 = new JObject() { 
                { "val1", -5.2585885403382 } 
            };
            var str1 = obj1.ToString();

            JObject obj2 = JObject.Parse(str1);
            var str2 = obj2.ToString();

            if (str1 != str2) throw new ApplicationException();
        }
```
[TestDoubleNewtonsoft.zip](https://github.com/JamesNK/Newtonsoft.Json/files/12451280/TestDoubleNewtonsoft.zip)

**Comment by elgonzo** [comment]
> Of course, the first example does pretty much that. Run in 64 bit it fails

Thanks for getting back. I tested again, and i found out why i did see str1 and str2 being equal -- in the legacy .NET Framework x64 runtime, the code behaves differently when running in debug mode compared to running in release mode. (And i apparently did my initial quick str1, str2 comparison test in release mode...)

Basically, building and running your example code in debug mode results in `str2` being:
```
{
  "val1": -5.2585885403382004
}
```
But building and running the same code in release mode, `str2` will be equal to `str1`: 
```
{
  "val1": -5.2585885403382
}
```

This problem can be traced back to Newtonsoft.Json using the `"R"` format specifier when placing floating point values into a json string:

https://github.com/JamesNK/Newtonsoft.Json/blob/0a2e291c0d9c0c7675d445703e51750363a549ef/Src/Newtonsoft.Json/JsonConvert.cs#L293-L296

This is what the official documentation has to say about the `"R"` specifier (https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings#round-trip-format-specifier-r):
> In .NET Framework and in .NET Core versions earlier than 3.0, the "R" format specifier **_fails to successfully round-trip [Double](https://learn.microsoft.com/en-us/dotnet/api/system.double) values in some cases._** For both [Double](https://learn.microsoft.com/en-us/dotnet/api/system.double) and [Single](https://learn.microsoft.com/en-us/dotnet/api/system.single) values, the "R" format specifier offers relatively poor performance. Instead, we recommend that you use the ["G17"](https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings#GFormatString) format specifier for [Double](https://learn.microsoft.com/en-us/dotnet/api/system.double) values and the ["G9"](https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings#GFormatString) format specifier to successfully round-trip [Single](https://learn.microsoft.com/en-us/dotnet/api/system.single) values.

And this made me remember that i made a comment to the same effect regarding another issue report about 2 years ago having the same root cause in Newtonsoft.Json using the `"R"` format specifier: https://github.com/JamesNK/Newtonsoft.Json/issues/2540 Unfortunately, this issue hasn't received any attention so far.

A possible workaround can be a custom JsonConverter for `double` values. This, however, will require the `JToken.ToString()` invocations to be altered, so i don't know whether this workaround would be feasible for you:
```C#
public class RoundtrippableDoubleConverter : JsonConverter<double>
{
    public override bool CanRead => false;

    public override double ReadJson(JsonReader reader, Type objectType, double existingValue, bool hasExistingValue, JsonSerializer serializer)
        => throw new NotSupportedException();

    public override void WriteJson(JsonWriter writer, double value, JsonSerializer serializer)
    {
        var str = value.ToString("G17");
        writer.WriteRawValue(str);
    }
}


static void Main(string[] args)
{
    var doubleConverter = new RoundtrippableDoubleConverter();
            
    JObject obj1 = new JObject() { 
        { "val1", -5.2585885403382 } 
    };
    
    //
    // Note the converter being passed to the ToString method.
    //
    var str1 = obj1.ToString(Newtonsoft.Json.Formatting.Indented, doubleConverter);

    JObject obj2 = JObject.Parse(str1);

    //
    // Note the converter being passed to the ToString method.
    //
    var str2 = obj2.ToString(Newtonsoft.Json.Formatting.Indented, doubleConverter);

    if (str1 != str2) throw new ApplicationException();
}
```

**Comment by RDUser-dev** [comment]
Hello @elgonzo, we had tried similar solutions in the past without success but your converter seems to solve the problem.
I try it inside the complete application and update you.
Thanks again

**Comment by RDUser-dev** [comment]
We applied the solution to a very complex project. Everything worked properly.
Thanks for your precious help @elgonzo .
I'm closing the issue
